keps/sig-node/0036-20190118-ForceEviction.md

```diff
@@ -0,0 +1,178 @@
+which blocks attach_detach controller to detach volumes. The current taint-based eviction actually
+cannot give any benefit in the sense that Pod can only be succesfully evicted after Kubelet comes
+back healthy, in which case there is no need of Pod eviction at all.
+
+
+### Goals
+
+This proposal is trying to investigate the failure handling issues for pods
+and their volumes with a goal of improving system reliability and failure recovery capabilities.
+Among different kinds of failures, node shut down is the main user case this proposal tries
+to address. In this case, we are sure that all containers are stopped, and all mounts are gone
+so that it is safe to take some recovery actions more aggresively without waiting for kubelet correspondence.  
+
+to delete the Pods from API server without waiting for kubelet's acknoledgement.
+Volumes can also be safely detached since mounts no longer exist.
+
+### Non-Goals
+
+This proposal is not trying to address the recovery issue for all types of failures. For example for network
+partition issue, because this is no source of truth about whether Pods are running or not, it is not possible
+for controller decide what recovery actions should take based on the node status.
+
+## Proposal
+
+Currently, Kubernetes already supports four types of taint effects:
+- NoSchedule: Pods that do not tolerate this taint are not scheduled on the node.
+- PreferNoSchedule: Kubernetes avoids scheduling Pods that do not tolerate this taint onto the node.
+- NoExecute: Pod is evicted from the node if it is already running on the node, and is not scheduled
+onto the node if it is not yet running on the node.
+
+We propose to add a new effect, TaintEffectForceEviction. For a Pod that cannot tolerant this taint,
+it will be force deleted from the node with zero grace period which means that Pod can be deleted immediately
+without waiting for Kubelet to responde. One important use case of this taint is node shut down case. If node
+controller or some external controller detact that node is powered off, the controller could add the following
+taint to the node.
+
+
+```
+taint := &v1.Taint{
+	Key:    nodeShutdown,
+	Value:  "true",
+	Effect: v1.TaintEffectForceEviction,
+}
+```
+
+After the node is tainted, node lifecycle controller will trigger to forcefully delete all the Pods from the node.
+If Kubelet comes back healthy in the middle of force eviction process, there might be race condition between kubelet
+starting Pods while they are being deleted since force deletion does not require kubelet coordination any more.
+To avoid race condition between controller and kubelet, kubelet also requires some additional handling when node
+is tainted which is explained in more details below.
+
+For node lifecycle controller,  we plan to add the following eviction logic:
+
+When node is marked as “NotReady” and tainted with “ForceEviction”. Taint manager will perform the following steps
```

Comment:
Maybe I missed it in the proposal, but it's not clear to me what adds this taint. Is this proposing to add it automatically based on some observed condition? Or is this a taint a cluster admin could manually add? If the former, most observed conditions could race with the node starting back up. If the latter, simply force deleting the pods seems simpler.

Reply:
It’s the cloud node_lifecycle that adds this taint. For the race condition, we ensure that when node comes backup it doesn’t start containers except static pods to ensure we aren’t detaching volumes that are still mounted.

Reply:
> It’s the cloud node_lifecycle that adds this taint. For the race condition, we ensure that when node comes backup it doesn’t start containers 

Can you describe the interlock between the node controller, the cloud provider API that ensures the "shutdown" state the controller observed from the cloud provider API is not stale? (e.g. the node has already started, and the kubelet has already started pods back up after the controller observed the shutdown indication in the cloud provider API but before adding the taint)

Reply:
If we taint the node but right after it came back, before the eviction starts we see if the node reported any heartbeat taking into account node-monitor-grace-period and pod eviction timeout. Since the kubelet won’t start any pods except static pods if it cannot connect to the apiserver, the eviction logic will assume that it is safe evict pods.

Reply:
This also suppose the kubelet will stop the containers on the node once it cannot update its lease (taking also into account the grace-period)